doc/SEMGREP_CORE_CONTRIBUTING.md

```diff
@@ -0,0 +1,535 @@
+
+Running (examples in Python):
+
+* To match a rule file against a target: `sc -config [your-rule].yaml [your-target].py -lang python`
+* To match a pattern against a target: `sc -f [your-pattern].sgrep [your-target].py -lang python`
+* To dump a pattern AST: `sc -dump_pattern [your-pattern].sgrep -lang python`
+* To dump a target AST: `sc -dump_ast [your-target].py -lang python` 
+* To dump a pattern Python AST: `pf -dump_python [your_pattern].sgrep -sgrep_mode -lang python`
+* To dump a target Python AST: `pf -dump_python [your_pattern].sgrep -lang python`
+
+Try it out: `sc -f tests/python/dots_stmts.sgrep tests/python/dots_stmts.py -lang python`
+
+## `semgrep-core` Overview
+
+### Entry Point
+
+The entry point to `semgrep-core` is `Main.ml`, in `semgrep-core/src/cli/`. This is where you add command-line arguments. It calls functions depending on the mode in which `semgrep-core` was invoked (`-config` for a yaml file, `-f` for a single pattern, etc.)
+
+When invoked by `semgrep`, `semgrep-core` is called by default wih `-config`. This corresponds to the function `semgrep-with-rules-file`, which in turn calls `semgrep-with-rules`. These functions will parse and then match the rule and targets.
+
+### Parsing
+
+`semgrep-core` uses external modules to parse code into augmented language-specific abstract syntax trees (ASTs). Though we call these ASTs, they additionally contain token information such as parentheses that are traditionally only present in concrete syntax trees (CSTs) so that we can output results in the correct range.
+
+When `semgrep-core` receives a rule or a target, it will first need to parse it. The functions that do this are located in `semgrep-core/src/parsing/`. 
+
+* If it reads a rule, it will go through `Parse_rule.ml`, which uses `Parse_pattern.ml` to parse the code-like portions of the rule
+* If it reads a target, it will go through `Parse_target.ml`
+
+Depending on the language, `Parse_pattern.ml` and `Parse_target.ml` will invoke parsers to parse the code. For example, if we have Java code, it will first be parsed into a Java-specific AST.
+
+### Converting to the Generic AST
+
+`semgrep-core` does not match based on the Java AST. It has a generic AST, defined in `AST_generic.ml` (in `semgrep-core/src/core/ast`), which all language-specific ASTs are converted to.
+
+The functions for this conversion are in either `semgrep-core/src/parsing/pfff/` or `semgrep-core/src/parsing/tree-sitter/`. They are named with the appropriate language in a consistent convention.
+
+### Matching
+
+The matching functions are contained in `semgrep-core/src/engine/` (e.g. `Match_rules.ml`, `Match_patterns.ml`) and `semgrep-core/src/matching/` (e.g. `Generic_vs_generic.ml`). There are several possible matchers to invoke
+
+* spacegrep (for generic mode)
+* regexp (to match by regexp instead of semgrep patterns)
+* comby (an experimental mode for C++)
```

Comment:
it's not just for C++, it's for lots of languages, and mostly targeted as the one not yet supported by the main mode.